proposals/lip-0045.md

```diff
@@ -1575,6 +1446,47 @@ genesisInteroperabilityStoreSchema = {
 
+#### Genesis State Initialization
+
+During the genesis state initialization stage of a genesis block `g`, the following steps are executed. If any step fails, the block is discarded and has no further effect.
+
+Let `genesisBlockAssetBytes` be the `data` bytes included in the genesis block assets for the Interoperability module and let `interoperabilityAsset = decode(genesisInteroperabilityStoreSchema, genesisBlockAssetBytes)`. Let `chainInfos = interoperabilityAsset.chainInfos`.
+
+##### Genesis Asset Verification
+
+* Within `chainInfos`, check that all entries have a different `chainID`.
+* For every `chainInfo` in `chainInfos`, properties `chainInfo.chainData`, `chainInfo.channelData` and `chainInfo.chainValidators` must be non-empty.
+* For every `chainInfo` in `chainInfos` with non-empty `chainInfo.terminatedOutbox`, the field `chainInfo.terminatedState` must also be non-empty.
+* For every `chainInfo` in `chainInfos` with `chainInfo.terminatedState.initialized == False`, the field `chainInfo.terminatedOutbox` must be empty, `chainInfo.terminatedState.stateRoot` must be equal to `EMPTY_HASH` and `chainInfo.terminatedState.mainchainStateRoot` must not be equal to `EMPTY_HASH`. Conversely, if `chainInfo.terminatedState.initialized == True`, then `chainInfo.terminatedState.stateRoot` must not be equal to `EMPTY_HASH` and `chainInfo.terminatedState.mainchainStateRoot` must be equal to `EMPTY_HASH`. todo: must the terminated outbox be non-empty in this case?
+* For each entry `chainInfo` in `chainInfos`, let `chainData = chainInfo.chainData`, then check:
+    * all `chainData.name` are unique.
+    * all `chainData.name` have character set `a-z0-9!@$&_.`.
+    * property `chainData.status` is in set `{CHAIN_STATUS_REGISTERED, CHAIN_STATUS_ACTIVE, CHAIN_STATUS_TERMINATED}`.
+* For each entry `chainInfo` in `chainInfos`, let `activeValidators = chainInfo.chainValidators.activeValidators` and let `certificateThreshold = chainInfo.chainValidators.certificateThreshold`, then check:
+    * `activeValidators` must have at least 1 element and at most `MAX_NUM_VALIDATORS` elements.
+    * `activeValidators` must be ordered lexicographically by `blsKey` property.
+    * All `blsKey` properties must be pairwise distinct.  
+    * Let `totalWeight` be the sum of the `bftWeight` property of every element in `activeValidators`. Then `totalWeight` has to be less than or equal to `MAX_UINT64`.
+    * Check that `totalWeight//3 + 1 <= certificateThreshold <= totalWeight`, where `//` indicates integer division.
+    * Check that the corresponding `validatorsHash` stored in `chainInfo.chainData.lastCertificate.validatorsHash` matches with the value computed from `activeValidators` and `certificateThreshold`.
+* On a sidechain, if `chainInfos` has an entry for another sidechain (`chainID != getMainchainID()`), then there must also exist an entry for the mainchain (`chainID == getMainchainID`) and `interoperabilityAsset.ownChainInfo` must be non-empty.
```

Comment:
I think we should still define the rules for the chainInfo.chainData entry for the mainchain, i.e. the fact that chainID and name must match.

Reply:
This is now done